### DISCUSSION TITLE
Recommended way to call methods on a Fabric component

### DISCUSSION THREAD

**Discussion by fbeccaceci** [reactwg/react-native-new-architecture#235] [Q&A]
So, i'm trying to create a Fabric component for the mapbox navigation sdk. I have my Fabric component that renders a map to the screen and works fine till there. Now let's say i wanna have a method that i can imperatively call from JS to move the map camera around, something like moveCameraTo(opts: CameraOptions), where CameraOptions will be a struct like this:
```typescript
interface CameraOptions {
  latitude: number;
  longitude: number;
  duration: number; // duration in ms the camera movement will take
  /* other fields omitted */
}
```

How would i call this method on my fabric component?

I tried implementing something with NativeCommands

```typescript
import type { HostComponent, ViewProps } from 'react-native';
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';
import codegenNativeCommands from 'react-native/Libraries/Utilities/codegenNativeCommands';

export interface NativeProps extends ViewProps {
  /* ... */
}

export default codegenNativeComponent<NativeProps>('MapboxNavigationView');

type ComponentType = HostComponent<NativeProps>;

interface NativeCommands {
  moveCameraTo: (viewRef: React.ElementRef<ComponentType>) => void;
}

export const Commands: NativeCommands = codegenNativeCommands<NativeCommands>({
  supportedCommands: ['moveCameraTo'],
});

```

And then i can invoke the command like this 

```typescript
Commands.moveCameraTo(mapViewRef.current);
```

But this does not work if i add the ```CameraOptions``` argument to the command, it throws out this error

```
Error: Unsupported param type for method "moveCameraTo", param "opts". Found TSInterfaceDeclaration
```
which is caused by codegen when running ```RCT_NEW_ARCH_ENABLED=1 pod install```

So my question is first of all how do you achieve this with the new architecture so fabric and turbo modules?

Also, would it be possibile to get back a promise from ```moveCameraTo``` that resolves when the camera movement is done? (Given that in the native side i can get a callback for the camera movement completion i just don't know how to pass it back to the js side)

I stumbled upon [this article ](https://www.devas.life/how-to-get-an-instance-of-the-fabric-view-component-on-react-native/) which talks about getting the reference to a component in the native side in the new architecture, i think i could get something to work this way so i can use a turbo module to call native methods which can have object arguments and use promises and then it will be the native code responsibility to find the correct component. 

But then i found https://github.com/reactwg/react-native-new-architecture/discussions/29#discussioncomment-2571486 where the author is saying that due to concurrent mode it is possibile that multiple version of the same view could exists so is the approach i have mentioned safe?

**Comment by vladimirivanoviliev** [comment]
I have the same questions as you:

1. Having a way to call methods on native component - seems that native commands are the solution for this, but haven't tried yet as this is the minor issue in our case (I have a way to workaround this).
2. Having a way to get the view instance and call methods to it -  from what I found so far it's no longer supported to get the native view instance. The method for getting the react tag is also deprecated (findNodeHandle). The approach in the linked article seems incorrect as it's accessing private properties, which might break without notice in future react versions. 

For me it's still hard to believe that no one have solution / guidance for this common case when working with third party native views.

I will keep you posted if I find something better of a solution.

**Reply by vladimirivanoviliev** [reply]
I implemented following solution - define separate turbo module that exposes the required methods. This module is singleton to which initialized views register and deregister. When the method on the module is called it get's the currently rendered view instance and calls the appropriate imperative method.

**Reply by vladimirivanoviliev** [reply]
Link to discussion I started in another thread [here](https://github.com/reactwg/react-native-new-architecture/discussions/8#discussioncomment-12725713).

**Reply by vladimirivanoviliev** [reply]
After the discussion it seems new recommended way is to use NativeCommands, however due to limited documentation and demos I will wait with my current solution till more info on these is available.
